### PR TITLE
avoid applying focus if dialog is not currentOverlay

### LINKED ISSUE #46: Multiple modal dialogs on screen throw 'Maximum call stack size exceeded' errors
_From @MoonDawg92 on November 16, 2015 16:33_

We have two separate paper-dialog windows and occasionally open them both at the same time. Whenever we focus on the topmost dialog, several errors are thrown as shown below.

error

We found this to only happen when both of the dialogs have the 'modal' attribute. Removing that attribute prevented the errors from being thrown, but we need it as a modal for our program.

Is this a known issue? Thanks!

Paper-Dialog Version 1.0.2
Google Chrome Version 46.0.2490.86 m

_Copied from original issue: PolymerElements/paper-dialog#69_

### PR DESCRIPTION
**Pull request by valdrinkoshi**
Fixes #46 by checking if the dialog is the currentOverlay before applying focus.

### CODE REVIEW

**Comment by valdrinkoshi** [comment]
@cdata, could you please review this? I tried to write some tests but I have problems in setting the focus to the elements, basically _onFocus won't get triggered on the test env.

File: paper-dialog-behavior.html [hunk 1]
```diff
@@ -205,32 +205,36 @@
     },
 
     _onFocus: function(event) {
-      if (this.modal) {
```
Comment (cdata) [inline paper-dialog-behavior.html]: Why did we stop caring about the `modal` property?

**Comment by cdata** [comment]
In order to trigger focus, a tabindex needs to be set on the thing that is being focused. Is that the issue?

File: paper-dialog-behavior.html (hunk 1 above)
Comment (valdrinkoshi) [inline paper-dialog-behavior.html]: `_onFocus` is used in `_boundOnFocus`, which is bound to body 'focus' only for modals (in `_onIronOverlayOpened`).

**Comment by valdrinkoshi** [comment]
No it's not. Each opened & modal `simple-dialog` will have an active listener on body click and backdropElement click. When we open both of the dialogs, these listeners are triggered for both, and they'll try to set the focus back to their `_lastFocusedElement` infinitely.

**Comment by valdrinkoshi** [comment]
@cdata, i realized that i was putting a spy on `_onFocus` instead of `_boundOnFocus`, and that was the whole issue for my tests :anguished: 
Now all should be green & ready for review

File: paper-dialog-behavior.html (hunk 1 above)
Comment (cdata) [inline paper-dialog-behavior.html]: What happens if `modal` changes while the dialog is open?

File: test/paper-dialog-behavior.html [hunk 2]
```diff
@@ -178,6 +197,42 @@ <h2 id="header">Dialog</h2>
...
+          var dialogs = fixture('multiple');
+          var spy = sinon.spy(dialogs[0], '_boundOnFocus');
+          runAfterOpen(dialogs[0], function() {
+            runAfterOpen(dialogs[1], function() {
+              var button = Polymer.dom(dialogs[1]).querySelector('[extra]');
+              button.click();
+              setTimeout(function() {
+                // Should not enter in an infinite loop.
+                assert.notEqual(spy.callCount, 0, 'focus changed at least once.');
+                // On IE focus changes twice.
```
Comment (cdata) [inline test/paper-dialog-behavior.html]: Why does it change twice in IE?

File: paper-dialog-behavior.html (hunk 1 above)
Comment (valdrinkoshi) [inline paper-dialog-behavior.html]: Should we handle that case? There are many other things would need to be fixed in that case, see here 
http://jsbin.com/neyudo/2/edit?html,output
Can you open an issue for that?

I'll restore the if check in here.

File: test/paper-dialog-behavior.html (hunk 2 above)
Comment (valdrinkoshi) [inline test/paper-dialog-behavior.html]: This is caused by the fact that we set [`useCapture`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) to `true` in here (the third param) https://github.com/PolymerElements/paper-dialog-behavior/blob/master/paper-dialog-behavior.html#L197

Here a quick example with just two buttons, also in chrome we see that focus callback is called twice http://jsbin.com/xobeqe/edit?html,output

Do you think it's safe to set `useCapture` to `false`?

File: test/paper-dialog-behavior.html [hunk 3]
```diff
@@ -178,6 +197,42 @@ <h2 id="header">Dialog</h2>
...
+          var dialogs = fixture('multiple');
+          var spy = sinon.spy(dialogs[0], '_boundOnFocus');
+          dialogs[1].addEventListener('iron-overlay-opened', function(){
+            var button = Polymer.dom(dialogs[1]).querySelector('[extra]');
+            button.click();
+            setTimeout(function() {
+              // Should not enter in an infinite loop.
+              assert.notEqual(spy.callCount, 0, 'focus changed at least once.');
+              // On IE focus changes twice.
+              assert.isTrue(spy.callCount <= 2, 'focus changed not more than twice.');
```
Comment (valdrinkoshi) [inline test/paper-dialog-behavior.html]: @cdata here we need to check twice because we set [`useCapture`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) to `true` in here (the third param) https://github.com/PolymerElements/paper-dialog-behavior/blob/master/paper-dialog-behavior.html#L197

Here a quick example with just two buttons, also in chrome we see that focus callback is called twice http://jsbin.com/makujo/3/edit?html,output

Do you think it's safe to set `useCapture` to `false`?

**Comment by valdrinkoshi** [comment]
@cdata feedback incorporated, PTAL

File: paper-dialog-behavior.html (hunk 1 above)
Comment (cdata) [inline paper-dialog-behavior.html]: Well, the reason I ask is that we have a change handler for when `modal` changes. This suggests that the current implementation handles changes of the `modal` property. We should probably not deviate from the current implementation. WDYT?

Comment (valdrinkoshi) [inline paper-dialog-behavior.html]: yes i agree and I've put back the check for modal.
There are still other issues related to the change of modal while the dialog is opened though, e.g. backdrop is not removed, listeners are not toggled. This stuff should be handled in a separate PR

File: paper-dialog-behavior.html [hunk 4]
```diff
@@ -205,32 +205,37 @@
...
-        }
-        if (target) {
-          if (target === document.body) {
-            if (this._lastFocusedElement) {
-              this._lastFocusedElement.focus();
-            } else {
-              this._focusNode.focus();
-            }
+      // If this is not modal or the top overlay, don't do anything.
+      if (!this.modal || this._manager.currentOverlay() !== this) {
```
Comment (valdrinkoshi) [inline paper-dialog-behavior.html]: @cdata here we still check for `this.modal`, if it is not true, we return

File: paper-dialog-behavior.html (hunk 1 above)
Comment (cdata) [inline paper-dialog-behavior.html]: SGTM

File: paper-dialog-behavior.html [hunk 5]
```diff
@@ -225,7 +225,7 @@
     },
 
     _onBackdropClick: function() {
-      if (this.modal) {
+      if (this.modal && this._manager.currentOverlay() === this) {
```
Comment (cdata) [inline paper-dialog-behavior.html]: :+1:

**Comment by cdata** [comment]
LGTM :+1:
